compiler/rustc_hir_analysis/src/constrained_generic_params.rs

```diff
@@ -43,26 +44,45 @@ pub fn parameters_for_impl<'tcx>(
 /// of parameters whose values are needed in order to constrain `ty` - these
 /// differ, with the latter being a superset, in the presence of projections.
 pub fn parameters_for<'tcx>(
+    tcx: TyCtxt<'tcx>,
     t: &impl TypeVisitable<TyCtxt<'tcx>>,
     include_nonconstraining: bool,
 ) -> Vec<Parameter> {
-    let mut collector = ParameterCollector { parameters: vec![], include_nonconstraining };
+    let mut collector =
+        ParameterCollector { tcx, parameters: vec![], include_nonconstraining, depth: 0 };
     t.visit_with(&mut collector);
     collector.parameters
 }
 
-struct ParameterCollector {
+struct ParameterCollector<'tcx> {
+    tcx: TyCtxt<'tcx>,
     parameters: Vec<Parameter>,
     include_nonconstraining: bool,
+    depth: usize,
 }
 
-impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for ParameterCollector {
+impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for ParameterCollector<'tcx> {
+    type BreakTy = ();
+
     fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
         match *t.kind() {
-            ty::Alias(..) if !self.include_nonconstraining => {
-                // projections are not injective
+            ty::Alias(ty::Projection | ty::Inherent | ty::Opaque, _)
+                if !self.include_nonconstraining =>
+            {
+                // Projections are not injective in general.
                 return ControlFlow::Continue(());
             }
+            ty::Alias(ty::Weak, alias) if !self.include_nonconstraining => {
+                if !self.tcx.recursion_limit().value_within_limit(self.depth) {
+                    return ControlFlow::Break(());
```

Comment:
We might want to `Continue` instead of `Break`, not sure.

Reply:
What's the difference?

Reply:
Only affects the diagnostics (“found unconstrained param” vs. “overflow”) and that only in certain cases. E.g., given:

```rs
#![feature(lazy_type_alias)]
#![allow(incomplete_features)]

trait Trait {}
impl<T> Trait for (Loop<T>, T) {}
type Loop<T> = Loop<T>;

fn main() {}
```

* With `Break`: ``the type parameter `T` is not constrained``
* With `Continue`: 2× `` overflow evaluating the requirement `Loop<T>` ``

Reply:
What's happening here (pretty sure): With `Break` we immediately give up, therefore don't see the bare `T`, consider `T` unconstrained due to the overflow and report a fatal “unconstrained” error. With `Continue` we – well – continue and see that `T` is indeed constrained. Hence we reach normalization & the overflow. So nothing major basically.

Reply:
I guess I like `Continue` better, lemme change it.